### PR TITLE
Change tracker docs

### LINKED ISSUE #505: New feature topic: Attach can track a graph of new and existing entities
Originally from #447.

### LINKED ISSUE #531: Add doc page for notification entities
Something like https://blog.oneunicorn.com/2016/11/16/notification-entities-in-ef-core-1-1/

### LINKED ISSUE #786: Clarify what entity states result from Add/Attach/Update on a graph
## Overview
In the [docs](https://docs.microsoft.com/en-us/ef/core/saving/related-data), it demonstrates how to add an entity with new related entities. What is the recommended method for adding a new entity that is related to an already existing entity?  

***Note***: I wasn't sure whether this issues is more relevant to the EF project itself, or the documentation. As it is more geared towards clarification of best practices, I decided to start here.

## Scenario
Say I am adding an `Item` that is dependent on a `Category`. My categories already exist, but the item that is being added does not. The entities have the following basic structure:

**Item.cs**
``` cs
public class Item
{
    public int Id { get; set; }
    public int CategoryId { get; set; }
    public string Name { get; set; }
    public Category Category { get; set; }
} 
```  

**Category.cs**
```  cs
public class Category
{
    public int Id { get; set; }
    public string Label { get; set; }
    public List<Item> Items { get; set; }
}
```  

The `Item` that I am providing has the following structure:  

``` js
{
    CategoryId: 1,
    Name: "Some Item",
    Category: {
        Id: 1,
        Label: "Some Category"
    }
}
```  

## Problem  
I have the following method for adding an `Item` to the database:  

``` cs
public static async Task AddItem(this AppDbContext db, Item item)
{
    if (item.Validate())
    {
        await db.Items.AddAsync(item);
        await db.SaveChangesAsync();
    }
}
```  

Whenever I try to save this way, it acts as though the related `Category` entity does not exist and tries to add a new instance to the database. In this case, it throws the following error:  

add-entity-error  

And the following is logged to the Output console in Visual Studio:  

``` cmd
EfAddFromApi.Web> Error Number:544,State:1,Class:16
EfAddFromApi.Web> fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
EfAddFromApi.Web>       Failed executing DbCommand (7ms) [Parameters=[@p0='?' (DbType = Int32), @p1='?' (Size = 4000)], CommandType='Text', CommandTimeout='30']
EfAddFromApi.Web>       SET NOCOUNT ON;
EfAddFromApi.Web>       INSERT INTO [Category] ([Id], [Label])
EfAddFromApi.Web>       VALUES (@p0, @p1);
EfAddFromApi.Web> System.Data.SqlClient.SqlException (0x80131904): Cannot insert explicit value for identity column in table 'Category' when IDENTITY_INSERT is set to OFF.

```

## Current Workaround  
I can get around this issue by setting any navigation properties to `null` before adding. When I do this, it works as expected.  

**Example**
``` cs
private static void ClearNavProps(this Item item)
{
    item.Category = null;
}

public static async Task AddItem(this AppDbContext db, Item item)
{
    if (item.Validate())
    {
        item.ClearNavProps();
        await db.Items.AddAsync(item);
        await db.SaveChangesAsync();
    }
}
```  

## Demo Application  
[Add to EF Core from Web API](https://github.com/JaimeStill/add-to-ef-from-api)

* `Item` is received from the `AddItem` method in the [`AppController`](https://github.com/JaimeStill/add-to-ef-from-api/blob/master/EfAddFromApi.Web/Controllers/AppController.cs)  
* `AddItem` extension method is defined in [`EntityExtensions`](https://github.com/JaimeStill/add-to-ef-from-api/blob/master/EfAddFromApi.Web/Extensions/EntityExtensions.cs#L30)

### LINKED ISSUE #815: Consider adding a more advanced many-to-many case.
For example:

```C#
public class DBUser
{
    public int Id { get; set; }

    public string Name { get; set; }

    public virtual ICollection<DBUserFriends> Friends { get; } = new List<DBUserFriends>();
    public virtual ICollection<DBUserBlockedUsers> BlockUsers { get; } = new List<DBUserBlockedUsers>();
}

public class DBUserFriends
{
    public int UserId { get; set; }
    public virtual DBUser User { get; set; }

    public int FriendId { get; set; }
    public virtual DBUser Friend { get; set; }
}

public class DBUserBlockedUsers
{
    public int UserId { get; set; }
    public virtual DBUser User { get; set; }

    public int BlockedUserId { get; set; }
    public virtual DBUser BlockedUser { get; set; }
}

public class BloggingContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder
            .UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Test;ConnectRetryCount=0");
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<DBUserFriends>(b =>
        {
            b.HasKey(e => new { e.UserId, e.FriendId });
            b.HasOne(e => e.User).WithMany(e => e.Friends);
            b.HasOne(e => e.Friend).WithMany().OnDelete(DeleteBehavior.ClientSetNull);
        });

        modelBuilder.Entity<DBUserBlockedUsers>(b =>
        {
            b.HasKey(e => new { e.UserId, e.BlockedUserId });
            b.HasOne(e => e.User).WithMany(e => e.BlockUsers);
            b.HasOne(e => e.BlockedUser).WithMany().OnDelete(DeleteBehavior.ClientSetNull);
        });
    }
}

public class Program
{
    public static void Main()
    {
        using (var context = new BloggingContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            var user1 = new DBUser();
            var user2 = new DBUser();
            var user3 = new DBUser();
            var user4 = new DBUser();
            
            user1.Friends.Add(new DBUserFriends { Friend = user2 });
            user1.Friends.Add(new DBUserFriends { Friend = user3 });
            user1.BlockUsers.Add(new DBUserBlockedUsers() { BlockedUser = user4 });
            context.Add(user1);

            context.SaveChanges();
        }

        using (var context = new BloggingContext())
        {
            foreach (var user in context.Set<DBUser>()
                .Include(e => e.Friends)
                .Include(e => e.BlockUsers)
                .ToList())
            {
                Console.WriteLine($"User {user.Id} has ");
                foreach (var friend in user.Friends.Select(e => e.Friend))
                {
                     Console.WriteLine($"  Friend {friend.Id}");
                }
                foreach (var blocked in user.BlockUsers.Select(e => e.BlockedUser))
                {
                    Console.WriteLine($"  Blocked {blocked.Id}");
                }
            }
        }
    }
}
```
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 24b2b50e-35f8-9e0b-8ce5-840e1ea816e6
* Version Independent ID: 9e0e6e2c-48fa-76c3-3bd9-813a6df22645
* Content: [Relationships - EF Core](https://docs.microsoft.com/en-us/ef/core/modeling/relationships)
* Content Source: [entity-framework/core/modeling/relationships.md](https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/modeling/relationships.md)
* Product: **entity-framework**
* GitHub Login: @rowanmiller
* Microsoft Alias: **divega**

### LINKED ISSUE #863: List of methods that call detectchanges?
would be nice to see here. Thanks.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7127791e-ded3-83b4-dbf0-ac332c91c10a
* Version Independent ID: 00d6bb6c-78cd-dfc3-83ad-20484fe7beb8
* Content: [ChangeTracker.DetectChanges Method (Microsoft.EntityFrameworkCore.ChangeTracking)](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.changetracking.changetracker.detectchanges?view=efcore-2.1#Microsoft_EntityFrameworkCore_ChangeTracking_ChangeTracker_DetectChanges)
* Content Source: [dotnet/xml/Microsoft.EntityFrameworkCore.ChangeTracking/ChangeTracker.xml](https://github.com/aspnet/EntityFramework.ApiDocs/blob/live/dotnet/xml/Microsoft.EntityFrameworkCore.ChangeTracking/ChangeTracker.xml)
* GitHub Login: @dotnet-bot

### LINKED ISSUE #1007: Clarify that fix-up to tracked entities is not materialization from the perspective of PropertyAccessMode
The documentation for UsePropertyAccessMode says that "By default, the backing field, if one is found by convention or has been specified, is used when new objects are constructed, typically when entities are queried from the database. Properties are used for all other accesses. Calling this method will change that behavior for all properties in the model as described in the PropertyAccessMode enum."

I think that even if the PropertyAccessMode is the default or is Field, during materialization the property setter will still be used for setting navigation properties as a result of EF's automatic fixing up of navigations to loaded entities. 

If this is true then it should be documented.

### LINKED ISSUE #1235: Clarify what happens in relationship fixup
In reference to this sentence 

> Note that blog will also be inserted into the database because it is a new entity that is referenced by the navigation property of an entity that is already tracked by the context (post).

What happens to the old `Blog` entity? Let's say that only the current post is in the collection of `Blog.Posts`, and I change the current post's blog to a new Blog entity. Will the old blog entity be removed from DB? Do I have to explicitly call `Remove` function?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 0f0fb981-4dec-7ff7-817b-6606f620b875
* Version Independent ID: c7517e7d-f011-a24c-53da-913771913998
* Content: [Saving Related Data - EF Core](https://docs.microsoft.com/en-us/ef/core/saving/related-data)
* Content Source: [entity-framework/core/saving/related-data.md](https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/saving/related-data.md)
* Product: **entity-framework**
* GitHub Login: @rowanmiller
* Microsoft Alias: **divega**

### LINKED ISSUE #1339: Clarify append-only semantics and short-lived context lifetimes
This issue tracker is for ***documentation***

For ***product*** issues, use https://github.com/aspnet/EntityFramework/issues

Relating to your tip "Entity Framework Core will automatically fix-up navigation properties to any other entities that were previously loaded into the context instance. So even if you don't explicitly include the data for a navigation property, the property may still be populated if some or all of the related entities were previously loaded."

How do I ensure that a query retrieves a navigation property from the database even if it is already loaded into the context instance?

### LINKED ISSUE #1398: Update guidance on using default values
See discussion here: https://github.com/aspnet/EntityFrameworkCore/issues/15070#issuecomment-477220843

* For hand-written code document:
  * Use client side default values. This ensures the defaults are consistent in the domain model.
  * Show how store-generated values work in the "normal" case--e.g. with integers
  * Show how to use option B for when both the normal way and client-side defaults are not appropriate, also mentioning C
  * Show how to get the default value into the migration if it needs to be there

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1f32ff95-761b-279a-f5a9-7fe859121356
* Version Independent ID: 63cc902d-8261-8185-d0a0-c48c627234b1
* Content: [Default Values - EF Core](https://docs.microsoft.com/en-us/ef/core/modeling/relational/default-values)
* Content Source: [entity-framework/core/modeling/relational/default-values.md](https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/modeling/relational/default-values.md)
* Product: **entity-framework**
* GitHub Login: @rowanmiller
* Microsoft Alias: **divega**

### LINKED ISSUE #1453: Add more examples for CRUD scenarios
This is great for getting data out but simply changing the referenced object and saving existing parent changes through the EF context fails with errors about "AssociationSet is in an 'Added' state, correspnding [Parent Class] must also. Please provide examples of correctly updating such properties through EF.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 24b2b50e-35f8-9e0b-8ce5-840e1ea816e6
* Version Independent ID: 9e0e6e2c-48fa-76c3-3bd9-813a6df22645
* Content: [Relationships - EF Core](https://docs.microsoft.com/en-us/ef/core/modeling/relationships)
* Content Source: [entity-framework/core/modeling/relationships.md](https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/modeling/relationships.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @rowanmiller
* Microsoft Alias: **divega**

### LINKED ISSUE #1489: Document state change differences when using generated key values
Let's say I have a `Blog` entity with a `BlogPost` collection. I have a bunch of `BlogPost` I want to update.

Instead of trying to update existing entities, I just `Clear()` the collection and add new entities with the same PKs than the removed entities. The `EntityState` of the new entities will be `Added`, but EF will correctly generate UPDATE sql statements and successfully save the changes.

At some point, it was detected that the `Added` entities were actually `Modified` versions of removed entities. I haven't found any information about this except a mention of "conflict resolution" in the article. Is there any information about this correction mechanism? Is this scenario supported or does it just work by accident?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b9826f28-46b2-e845-f1bf-5cbc2d940208
* Version Independent ID: 9c4a46cf-daa2-0e14-7ff1-c195db237cd0
* Content: [Disconnected Entities - EF Core](https://docs.microsoft.com/en-us/ef/core/saving/disconnected-entities)
* Content Source: [entity-framework/core/saving/disconnected-entities.md](https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/saving/disconnected-entities.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @ajcvickers
* Microsoft Alias: **avickers**

### LINKED ISSUE #1541: Expand change tracking topic
We should cover things like automatic change detection, snapshot vs. notification entities, blobs, guidance on how to implement Equals() and GetHashCode(), etc.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d3527ab8-2b07-a059-bd23-5f0700cfa0fc
* Version Independent ID: 1e317ee0-fe29-b37f-43ec-a0ac64e17db9
* Content: [Saving Data - EF Core](https://docs.microsoft.com/en-us/ef/core/saving/)
* Content Source: [entity-framework/core/saving/index.md](https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/saving/index.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @rowanmiller
* Microsoft Alias: **divega**

### LINKED ISSUE #2076: Document change tracking proxies
See https://github.com/dotnet/efcore/issues/10949

### LINKED ISSUE #2105: Clarify how generated key sentinel and temporary values work
For example:
* The CLR default acts as a sentinel...
* But only when using generated key values.

See https://github.com/dotnet/efcore/issues/19829

### LINKED ISSUE #2331: Document Find
For a beginner this is impossible to follow without examples. In some cases, isn't the entity also the object we want to search?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 24ff656e-4936-b37c-e1cc-1824aa148ccd
* Version Independent ID: b18f904f-d60c-955b-5aee-b739ffd8fe9f
* Content: [DbContext.FindAsync Method (Microsoft.EntityFrameworkCore)](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.findasync?view=efcore-3.1)
* Content Source: [dotnet/xml/Microsoft.EntityFrameworkCore/DbContext.xml](https://github.com/aspnet/EntityFramework.ApiDocs/blob/live/dotnet/xml/Microsoft.EntityFrameworkCore/DbContext.xml)
* Product: **entity-framework-core**
* GitHub Login: @dotnet-bot
* Microsoft Alias: **divega**

### LINKED ISSUE #2524: Document ChangeTracker.Clear()

### LINKED ISSUE #2525: Document nullable backing field pattern for store-generated defaults
https://github.com/dotnet/efcore/issues/15182

### LINKED ISSUE #2707: Document many-to-many change tracking
Including disconnected scenarios.

### LINKED ISSUE #2725: Document DbSet.Local and related concepts
This issue tracker is for ***documentation***

For ***product*** issues, use https://github.com/aspnet/EntityFramework/issues

### PR DESCRIPTION
**Pull request by ajcvickers**
Change tracker docs

Fixes #2725
Fixes #2707
Fixes #2525
Fixes #2524
Fixes #2331
Fixes #2105
Fixes #2076
Fixes #1541
Fixes #1489
Fixes #1453
Fixes #1398
Fixes #1339
Fixes #1235
Fixes #1007
Fixes #863 
Fixes #815
Fixes #786 
Fixes #531 
Fixes #505

### CODE REVIEW

**Comment by ajcvickers** [comment]
@dotnet/efteam The "estimated time to read" for all this is just over 2 hours. But you guys are fast readers, right? 😉

File: entity-framework/core/change-tracking/change-detection.md [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+- When using <xref:Microsoft.EntityFrameworkCore.ChangeTracking.EntityEntry> methods such as `Property`, `Collection`, `Reference` or `Member` to ensure property modifications, current values, etc. are up-to-date.
+- When an dependent/child entity is going to be deleted because a required relationship has been severed. This detects when an entity should not be deleted because it has been re-parented.
+
+Local detection of changes for a single entity can be triggered explicitly by calling <xref:Microsoft.EntityFrameworkCore.ChangeTracking.EntityEntry.DetectChanges?displayProperty=nameWithType>.
+
+> [!NOTE]
+> Local detect changes can miss some changes that a full detection would fine. This happens when cascading actions resulting from undetected changes to other entities have an impact on the entity in question. In such situations the application may need to force a full scan of all entities by explicitly calling <xref:Microsoft.EntityFrameworkCore.ChangeTracking.ChangeTracker.DetectChanges?displayProperty=nameWithType>.
```
Comment (JeremyLikness) [inline entity-framework/core/change-tracking/change-detection.md:148]: fine -> find

File: entity-framework/core/change-tracking/change-detection.md [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+
+### Disabling automatic change detection
+
+The performance of detecting changes is not a bottleneck for most applications. However, detecting changes can become a performance problem for some applications that track thousands of entities. For this reason the automatic detection of changes can be disabled using <xref:Microsoft.EntityFrameworkCore.ChangeTracking.ChangeTracker.AutoDetectChangesEnabled?displayProperty=nameWithType>. For example, consider processing join entities in a many-to-many relationship with payloads:
```
Comment (JeremyLikness) [inline entity-framework/core/change-tracking/change-detection.md:152]: thousands of entities ... or more.

File: entity-framework/core/change-tracking/identity-resolution.md [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
+            }
+-->
+[!code-csharp[Identity_Resolution_in_EF_Core_1](../../../samples/core/ChangeTracking/IdentityResolutionInEFCore/IdentityResolutionSamples.cs?name=Identity_Resolution_in_EF_Core_1)]
+
+Running this code results in the following exception:
+
+> System.InvalidOperationException: The instance of entity type 'Blog' cannot be tracked because another instance with the key value '{Id: 1}' is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached.
```
Comment (JeremyLikness) [inline entity-framework/core/change-tracking/identity-resolution.md:42]: Should this be monospaced, i.e. surrounded with backticks?

File: entity-framework/core/change-tracking/miscellaneous.md [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+
+Other similar methods like `Update`, `Attach`, and `Remove` do not have async overloads because they never generate new key values, and hence never need to access the database.
+
+## AddRange, UpdateRange, AttachRange, and RemoveRange
+
+<xref:Microsoft.EntityFrameworkCore.DbSet%601> and <xref:Microsoft.EntityFrameworkCore.DbContext> provide alternate versions of `Add`, `Update`, `Attach`, and `Remove` that accept multiple instances in a single call. These methods are called `AddRange`, `UpdateRange`, `AttachRange`, and `RemoveRange` respectively.
+
+These methods are provided as a convenience. Using a "range" method has the same functionality as multiple calls to the equivalent non-range method. There is no significant performance different between the two approaches.
```
Comment (JeremyLikness) [inline entity-framework/core/change-tracking/miscellaneous.md:31]: performance different -> performance difference

**Review by JeremyLikness — APPROVED** [review]
Some minor questions and fixes.

File: entity-framework/core/change-tracking/explicit-tracking.md [hunk 5]
```diff
@@ -0,0 +1,991 @@
...
+
+Entity Framework Core (EF Core) change tracking works best when the same <xref:Microsoft.EntityFrameworkCore.DbContext> instance is used to both query for entities and update them by calling <xref:Microsoft.EntityFrameworkCore.DbContext.SaveChanges%2A>. This is because EF Core automatically tracks the state of queried entities and then detects any changes made to these entities when SaveChanges is called. This approach is covered in [Change Tracking in EF Core](xref:core/change-tracking/index).
+
+> [!TIP]
+> This document assumes that entity states and the basics of EF Core change tracking are understood. See [Change Tracking in EF Core](xref:core/change-tracking/index) for more information on these topics.
+
+> [!TIP]  
+> You can run and debug into all the code in this document by [downloading the sample code from GitHub](https://github.com/dotnet/EntityFramework.Docs/tree/master/samples/core/ChangeTracking/ChangeTrackingInEFCore).
+
+> [!TIP]
```
[the remark this answers is not in the document]
Comment (ajcvickers) [reply entity-framework/core/change-tracking/explicit-tracking.md:21]: It looks like we pretty consistently use Tip for this. However, not against changing it everywhere if Note is more appropriate. /cc @dotnet/efteam

File: entity-framework/core/change-tracking/identity-resolution.md (hunk 3 above)
Comment (ajcvickers) [reply entity-framework/core/change-tracking/identity-resolution.md:42]: I think it reads better like this, but happy to change if the team agrees.
/cc @dotnet/efteam

**Pull request merged**